sxt/multiexp/bucket_method/combination_kernel.t.cc

```diff
@@ -0,0 +1,114 @@
+/** Proofs GPU - Space and Time's cryptographic proof algorithms on the CPU and GPU.
+ *
+ * Copyright 2023-present Space and Time Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "sxt/multiexp/bucket_method/combination_kernel.h"
+
+#include "sxt/base/curve/example_element.h"
+#include "sxt/base/device/synchronization.h"
+#include "sxt/base/test/unit_test.h"
+#include "sxt/memory/management/managed_array.h"
+#include "sxt/memory/resource/managed_device_resource.h"
+
+using namespace sxt;
+using namespace sxt::mtxbk;
+
+TEST_CASE("we can combine partial bucket sums") {
+  memmg::managed_array<bascrv::element97> partial_bucket_sums(memr::get_managed_device_resource());
+  memmg::managed_array<bascrv::element97> bucket_sums(memr::get_managed_device_resource());
+
+  SECTION("we handle the case of a single bucket") {
+    partial_bucket_sums = {12u};
+    bucket_sums = {0u};
+    combine_partial_bucket_sums<<<1, 1>>>(bucket_sums.data(), partial_bucket_sums.data(), 1);
+    basdv::synchronize_device();
+    REQUIRE(bucket_sums[0] == 12u);
+  }
+
+  SECTION("we handle a bucket with two sums") {
+    partial_bucket_sums = {12u, 45u};
+    bucket_sums = {0u};
+    combine_partial_bucket_sums<<<1, 1>>>(bucket_sums.data(), partial_bucket_sums.data(), 2);
+    basdv::synchronize_device();
+    REQUIRE(bucket_sums[0] == 57u);
+  }
+
+  SECTION("we handle a bucket group size > 1") {
+    partial_bucket_sums = {12u, 45u, 7u, 3u};
+    bucket_sums = {0, 0};
+    combine_partial_bucket_sums<<<1, 2>>>(bucket_sums.data(), partial_bucket_sums.data(), 2);
+    basdv::synchronize_device();
+    REQUIRE(bucket_sums[0] == 19u);
+    REQUIRE(bucket_sums[1] == 48u);
+  }
+
+  SECTION("we handle more than one bucket group") {
+    partial_bucket_sums = {12u, 45u, 7u, 3u};
+    bucket_sums = {0, 0};
+    combine_partial_bucket_sums<<<2, 1>>>(bucket_sums.data(), partial_bucket_sums.data(), 2);
```

Comment:
> Is there a case where partial_bucket_sums will be used after calling combine_partial_bucket_sums?

This will make a bit more sense once the other pieces are in. But, no, it's deallocated after combine_partial_bucket_sums is finished and it's allowed to be used as workspace memory for inplace addition.